### PR TITLE
add rapids-get-pr-wheel-artifact for quickly downloading cpp wheels

### PR DESCRIPTION
**Pull request by msarahan**
This idea comes from a slack DM with Viyas:

> OK, so the first piece I'm thinking of is artifact download. If you look at the raft PR, we have [snippets like this](https://github.com/rapidsai/raft/pull/2251/files#diff-616eb8e64286a9a59a2cd9c77d817fba66ce78ecfdde635db9f4028b74349585R34):
artifact_name=$(RAPIDS_PY_WHEEL_NAME="librmm_${RAPIDS_PY_CUDA_SUFFIX}" RAPIDS_REPOSITORY=rmm RAPIDS_PY_VERSION="3.11" rapids-package-name wheel_python)
commit=$(git ls-remote https://github.com/rapidsai/rmm.git refs/heads/pull-request/1512 | cut -c1-7)
librmm_wheelhouse=$(rapids-get-artifact "ci/rmm/pull-request/1512/${commit}/${artifact_name}")

> The snippet above demonstrates two issues:
There isn't an easy way to download arbitrary PR wheel artifacts. However, [there is such a tool for conda artifacts](https://github.com/rapidsai/gha-tools/blob/main/tools/rapids-get-pr-conda-artifact). That was added in https://github.com/rapidsai/gha-tools/pull/85. What we want is to generalize this tool to work for wheels.
The librmm wheel should not be specific to a particular Python version because it's a pure C++ wheel. However, our current naming scheme on [downloads.rapids.ai](http://downloads.rapids.ai/) doesn't know that. I'm currently hacking it by specifying the RAPIDS_PY_VERSION="3.11". This is working because [I'm restricting the C++ wheel build CI job to only run with Python 3.11](https://github.com/rapidsai/raft/pull/2251/files#diff-1eb4e5fd5611777d4e597ef299a1cb5ba8050c28a2dabbd4fbc56205d69e5ddaR82). There was [another recent gha-tools PR](https://github.com/rapidsai/gha-tools/pull/96) that added support for pure Python wheels. What we want is to generalize this tool to also work for C++ wheels. The main difference is that C++ wheels have to care about arch, whereas pure Python wheels do not

> I think a good starting point would be to make PRs to gha-tools to address the above issues. Does that make sense?

At the time that I'm submitting this, I haven't yet found examples of the wheels_cpp files. Vyas is working on fleshing those out, and I will be taking over that work from him once he gets it on the right track.

### CODE REVIEW

**Comment by vyasr** [comment]
Note that we may not update every repo that's using these tools to the new wheels strategy, but we will at least have to ensure that we update them for the breaking changes here.

**Comment by msarahan** [comment]
On second thought, the breaking change here may not be the best approach. I thought it would be a helpful guide for which repos need fixing to add wheel support. If it's going to disrupt things with a lot of repos for no change to those repos, then I think it might be better to preserve the wheel_python default behavior.

Maybe something like:

```
# default assumes wheel_python. If manually specified, do that instead.
pkg_name="$(rapids-package-name wheel_python)"

if [ "$1" = "cpp" ] || [ "$1" = "python" ]; then
	pkg_type=$1
	pkg_name="$(rapids-package-name "wheel_${pkg_type}")"
	# remove pkg_type from args because we handle it in this script
	shift;
fi
```

**Review by vyasr — COMMENTED** [review]
Yikes I reviewed this a week ago but apparently I never submitted. Sorry about that.

File: tools/rapids-download-wheels-from-s3
Comment (bdice) [reply tools/rapids-download-wheels-from-s3:9]: I'm fine with the code here. We can have a follow-up to clean up all uses of the script so they declare the `pkg_type`.

File: tools/rapids-package-name
Comment (bdice) [inline tools/rapids-package-name:73]: Let's consolidate this with `RAPIDS_PY_WHEEL_PURE` because it does basically the same thing.

"Pure" wheels can be either _just_ Python code (no extension modules unless they're abi3 or similar) or _just_ bundles of shared libraries that have no specific need for a Python minor version. Examples of the former include dask-cudf. Examples of the latter include RAPIDS C++ wheels or CUDA wheels.

Also just to be clear, we don't need a generic Python version string like `_py3` in the artifact name, it is already part of the wheel name contained in the artifact. All that we care is that artifact names are self-consistent and can be identified by human readers.

File: tools/rapids-download-wheels-from-s3
Comment (ajschmidt8) [reply tools/rapids-download-wheels-from-s3:9]: To be honest I'm hoping we can remove a lot of these scripts entirely pending the outcome of https://github.com/rapidsai/ops/issues/2982.

If that issue ends up being rolled out, I think that would be a good time to rethink some of the interfaces for any of the helper scripts that we do keep around.

File: tools/rapids-get-pr-wheel-artifact
Comment (ajschmidt8) [reply tools/rapids-get-pr-wheel-artifact:33]: > @ajschmidt8 does CI have the necessary credentials to make this request over SSH rather than HTTPS?

Nope. Good catch. Private repo failures will have to be handled appropriately here.

**Comment by vyasr** [comment]
It would also be good to [update the docs](https://docs.rapids.ai/resources/github-actions/#using-conda-ci-artifacts-in-other-prs) showing how this works once we finalize and merge this tool.

File: tools/rapids-get-pr-wheel-artifact
Comment (vyasr) [reply tools/rapids-get-pr-wheel-artifact:33]: Maybe we should do that for the conda tool as well in this PR then to keep things aligned.

Comment (msarahan) [reply tools/rapids-get-pr-wheel-artifact:33]: This is copied frrom the conda tool. I'd like to separate this issue of private repo support in this command from the effort of adding equivalency between conda and wheels.

Would it be OK with you (plural) to punt on this?

File: tools/rapids-package-name
Comment (msarahan) [reply tools/rapids-package-name:73]: seems like I should just removed this bit completely. Thanks for the simplification.

File: tools/rapids-get-pr-wheel-artifact
Comment (ajschmidt8) [reply tools/rapids-get-pr-wheel-artifact:33]: Yup, punt away

Comment (msarahan) [reply tools/rapids-get-pr-wheel-artifact:33]: https://github.com/rapidsai/gha-tools/issues/103

File: tools/rapids-package-name
Comment (bdice) [inline tools/rapids-package-name:34]: It doesn't look like we append the CUDA version for `wheel_python`. Is that because we're using wheel names that include the suffix like `-cu12`? We should probably mirror that behavior for `wheel_cpp`.

**Review by ajschmidt8 — APPROVED** [review]
No questions/objections from me.

Approved pending the resolution of Brad's questions

**Pull request merged**
